zcash-hackworks/zfaucet
# Ben/docker compose initial changes

Some small tweaks to disconnect the services to enable containerization.

Currently requires https://github.com/zcash-hackworks/pyZcash/pull/1 and a `.zcash` and `.zcash-params` so we're not waiting for network sync on every change.

## Steps to test:
```
git clone git@github.com:benzcash/zfaucet.git test-zfaucet
cd test-zfaucet/
git checkout ben/docker-compose-initial-changes
git clone git@github.com:benzcash/pyZcash.git lib/pyZcash
cd pyZcash
git checkout ben/enable-rpc-params-in-env
cd ..
cp .env-example .env
<EDIT .env>
docker-compose up -d --build
```

For the `edit .env`, step most setting are self explanatory. 

The  `ZCASHD_DATADIR` and `ZCASHD_PARMDIR` should point to an existing zcashd location, make sure to stop your existing process!

Visit http://127.0.0.1:8080 to access the site.

It's possible the `zfaucet` container failed to start because `zcashd` isn't ready.

```
$ docker-compose ps
         Name                       Command                      State                    Ports          
---------------------------------------------------------------------------------------------------------
test-zfaucet_db_1        docker-entrypoint.sh postgres   Up                      0.0.0.0:5432->5432/tcp  
test-zfaucet_nginx_1     nginx -g daemon off;            Up                      0.0.0.0:8080->80/tcp    
test-zfaucet_zcashd_1    zcashd                          Up (health: starting)   0.0.0.0:18232->18232/tcp
test-zfaucet_zfaucet_1   docker/zfaucet/entrypoint.sh    Exit 1 
```

Just start it up

```
$ docker-compose up -d zfaucet
test-zfaucet_db_1 is up-to-date
test-zfaucet_zcashd_1 is up-to-date
Starting test-zfaucet_zfaucet_1 ... done
$ docker-compose ps
         Name                       Command                      State                    Ports          
---------------------------------------------------------------------------------------------------------
test-zfaucet_db_1        docker-entrypoint.sh postgres   Up                      0.0.0.0:5432->5432/tcp  
test-zfaucet_nginx_1     nginx -g daemon off;            Up                      0.0.0.0:8080->80/tcp    
test-zfaucet_zcashd_1    zcashd                          Up (health: starting)   0.0.0.0:18232->18232/tcp
test-zfaucet_zfaucet_1   docker/zfaucet/entrypoint.sh    Up                      0.0.0.0:8000->8000/tcp
```


Todo;
- healthcheck on containers
- fail2ban with Docker :thinking: 
- cert mapping for production
